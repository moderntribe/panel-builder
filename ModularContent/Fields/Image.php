<?php

namespace ModularContent\Fields;
use ModularContent\AdminPreCache;
use ModularContent\Panel;

/**
 * Class Image
 *
 * @package ModularContent\Fields
 *
 * An image field.
 */
class Image extends Field {

	protected $default = 0;
	protected $layout  = 'compact';

	protected $default_mime_types = [
		'image/svg',
		'image/svg+xml',
		'image/jpeg',
		'image/gif',
		'image/png',
		'image/bmp',
		'image/tiff',
		'image/x-icon',
	];

	protected $allowed_image_mime_types;

	/**
	 * @param array $args
	 */
	public function __construct( $args = array() ) {
		$this->check_layout( $args );

		$this->defaults[ 'strings' ] = [
			'button.remove' => __( 'Remove', 'modular-content' ),
			'button.select' => __( 'Select Files', 'modular-content' ),
		];
		$this->defaults['allowed_image_mime_types'] = isset( $args['allowed_image_mime_types'] ) ? $args['allowed_image_mime_types'] : apply_filters( 'panels_default_allowed_mime_types', $this->default_mime_types );
		$this->defaults['layout'] = $this->layout;
		parent::__construct($args);
	}

	protected function check_layout( $args ) {
		if ( isset( $args['layout'] ) && $args['layout'] !== 'compact' && $args['layout'] !== 'full' && $args['layout'] !== 'inline' ) {
			throw new \LogicException( 'Layout argument can only be "compact" or "full".' );
		}
	}

	public function get_vars_for_api( $data, $panel ) {

		$all_sizes_data = [ ];

		// Full is the only guaranteed size, so it's going to be our default
		$size_data = wp_get_attachment_image_src( $data, 'full' );

		// Something went wrong. Most likely the attachment was deleted.
		if ( $size_data === false ) {
			return new \stdClass;
		}

		$attachment = get_post( $data );

		$return_data = [
			'url'         => $size_data[0],
			'width'       => $size_data[1],
			'height'      => $size_data[2],
			'title'       => $attachment->post_title,
			'alt'         => get_post_meta( $attachment->ID, '_wp_attachment_image_alt', true ),
			'description' => $attachment->post_content,
			'caption'     => $attachment->post_excerpt,
		];

		// Set all the other sizes

		foreach ( get_intermediate_image_sizes() as $size ) {

			if ( $size === 'full' ) {
				continue;
			}

			$size_data = wp_get_attachment_image_src( $data, $size );

			if ( $size_data === false ) {
				continue;
			}

			$all_sizes_data[ $size ] = [
				'url'    => $size_data[0],
				'width'  => $size_data[1],
				'height' => $size_data[2],
			];
		}

		$return_data['sizes'] = $all_sizes_data;

		$return_object = new \stdClass();
		foreach ( $return_data as $key => $value ) {
			$return_object->$key = $value;
		}

		$return_object = apply_filters( 'panels_field_vars_for_api', $return_object, $data, $this, $panel );

		return $return_object;
	}

	public function get_blueprint() {
		$blueprint = parent::get_blueprint();
		$blueprint['allowed_image_mime_types'] = $this->allowed_image_mime_types;
		$blueprint['layout'] = $this->layout;
		return $blueprint;
	}

	/**
	 * Add data relevant to this field to the precache
	 *
	 * @param mixed         $data
	 * @param AdminPreCache $cache
	 *
	 * @return void
	 */
	public function precache( $data, AdminPreCache $cache ) {
		if ( $data ) {
			$cache->add_image( $data );
		}
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * @param mixed $data
	 * @return int
	 */
	public function prepare_data_for_save( $data ) {
		return (int) parent::prepare_data_for_save( $data );
	}
}
