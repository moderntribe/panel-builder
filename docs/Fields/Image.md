## Image

An image field.

The image is stored in the field as an attachment ID.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |

### Example

```php
$field = new Image( array(
    'label' => __('Featured Image'),
    'name' => 'featured-image',
    'description' => __( 'An image to feature' ),
) );
```
