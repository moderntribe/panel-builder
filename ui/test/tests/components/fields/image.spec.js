/* global expect b:true */

import React from 'react';
import { shallow } from 'enzyme';
import Image from 'components/fields/image';
import renderer from 'react-test-renderer';

describe('Image field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Image />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	const wrapper = shallow(<Image />);

	it('renders as a <div>', () => {
		expect(wrapper.type()).toEqual('div');
	});
});
