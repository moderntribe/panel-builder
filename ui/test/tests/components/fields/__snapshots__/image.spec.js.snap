exports[`Image field renders correctly 1`] = `
<div
  className="field panel-field">
  <label
    className="label panel-field-label">
    
  </label>
  <div
    className="container">
    <div
      className="current-file currentUploadedFile">
      <div
        className="imagePreview">
        <img
          alt=""
          onClick={[Function]}
          src="" />
      </div>
      <p
        className="removeButtonContainer">
        <button
          className="panel-button secondary full"
          data-id=""
          data-loading={false}
          disabled={false}
          onClick={[Function]}
          title=""
          type="button">
          <span>
            
          </span>
        </button>
      </p>
    </div>
    <div
      className="file-uploader uploaderOpen uploaderSection">
      <button
        className="panel-button primary full attachment_helper_library_button"
        data-id=""
        data-loading={false}
        disabled={false}
        onClick={[Function]}
        title=""
        type="button">
        <span>
          
        </span>
      </button>
    </div>
  </div>
</div>
`;
